File: golang/eval/validator.go
package eval

import (
	"bytes"
	_ "embed"
	"fmt"
	"github.com/goccy/go-yaml"
	"github.com/santhosh-tekuri/jsonschema/v6"
)

type Validator interface {
	Validate(*ValidateInput) *ValidateOutput
}

//go:embed schema.yaml
var schema string

func NewValidator() Validator {
	b, err := yaml.YAMLToJSON([]byte(schema))
	if err != nil {
		panic(fmt.Sprintf("fail to convert yaml to json: %+v", err))
	}

	v, err := jsonschema.UnmarshalJSON(bytes.NewBuffer(b))
	if err != nil {
		panic(fmt.Sprintf("fail to decode json: %+v", err))
	}

	c := jsonschema.NewCompiler()
	if err := c.AddResource("https://github.com/Jumpaku/eyamluate", v); err != nil {
		panic(fmt.Sprintf("fail to add schema: %+v", err))
	}

	return &validator{schema: c.MustCompile("https://github.com/Jumpaku/eyamluate")}
}

type validator struct {
	schema *jsonschema.Schema
}

var _ Validator = &validator{}

func (v *validator) Validate(input *ValidateInput) *ValidateOutput {
	b, err := yaml.YAMLToJSON([]byte(input.Source))
	if err != nil {
		return &ValidateOutput{
			Status:       ValidateOutput_YAML_ERROR,
			ErrorMessage: fmt.Sprintf("fail to convert yaml to json: %+v", err),
		}
	}

	sourceJSON, err := jsonschema.UnmarshalJSON(bytes.NewBuffer(b))
	if err != nil {
		return &ValidateOutput{
			Status:       ValidateOutput_YAML_ERROR,
			ErrorMessage: fmt.Sprintf("fail to decode json: %+v", err),
		}
	}

	if err := v.schema.Validate(sourceJSON); err != nil {
		vo := &ValidateOutput{
			Status:       ValidateOutput_VALIDATION_ERROR,
			ErrorMessage: fmt.Sprintf("validation error: %#+v", err),
		}
		return vo
	}

	return &ValidateOutput{Status: ValidateOutput_OK}
}

File: dart/lib/eval/validator.dart
import 'dart:convert';

import 'package:embed_annotation/embed_annotation.dart';
import 'package:eyamluate/eval/eval.dart';
import 'package:json_schema/json_schema.dart';
import 'package:yaml/yaml.dart';

part 'validator.g.dart';

@EmbedBinary("../../../schema/eyamluate.schema.yaml")
const languageSchemaYAMLBytes = _$languageSchemaYAMLBytes;

class Validator {
  ValidateOutput validate(ValidateInput input) {
    final csvArmorSchemaYAML = utf8.decode(languageSchemaYAMLBytes);
    final schema = jsonEncode(loadYaml(csvArmorSchemaYAML));
    final validator = JsonSchema.create(schema);

    final dynamic v;
    try {
      v = loadYaml(input.source);
    } catch (e) {
      return ValidateOutput(
        status: ValidateOutput_Status.YAML_ERROR,
        errorMessage: e.toString(),
      );
    }

    final r = validator.validate(v);
    if (!r.isValid) {
      return ValidateOutput(
        status: ValidateOutput_Status.VALIDATION_ERROR,
        errorMessage: r.errors.toString(),
      );
    }

    return ValidateOutput(status: ValidateOutput_Status.OK);
  }
}
